Name: <redacted> <redacted> <redacted>
GitHub repo url: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>
commit id: 25ff7d81703b932d6f52568a5f827946a2fa4493

In [99]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Part 1: Load the dataset

In [100]:
# Load the dataset (load remotely or locally)
url = './iris.csv'
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv(url)
print(df.head(15))
print(df.info())

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This dataset contains information on various measurements of flowers from three species of Iris: setosa, versicolor, and virginica.

These are the columns in the dataset that are the features:
1. `sepal_length`: The length of the sepal (cm).
2. `sepal_width`: The width of the sepal (cm).
3. `petal_length`: The length of the petal (cm).
4. `petal_width`: The width of the petal (cm).

The `species` column is the label. 

The label (species name) in the dataset maps to an actual iris species class:
- Label `setosa` → Class *Iris setosa*
- Label `versicolor` → Class *Iris versicolor*
- Label `virginica`→ Class *Iris virginica*

In this classification task, the goal is to use the features to train a model that can predict the species class (label) for new flowers based on their measurements.




# Part 2: Split the dataset into train and test

In [101]:
# Take the dataset and split it into our features (X) and label (y)
X = df.drop(columns=['species'])
y = df['species']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=35)

# Part 3: Logistic Regression

In [102]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
# ii. For a sample datapoint, predict the probabilities for each possible class
model.fit(X_train, y_train)
sample = X_test.iloc[0]
print(model.predict([sample]))
print(model.predict_proba([sample]))
# iii. Report on the score for Logistic regression model, what does the score measure?
print(model.score(X_test, y_test))
# iv. Extract the coefficents and intercepts for the boundary line(s)
print(model.coef_)
print(model.intercept_)

['versicolor']
[[0.02341215 0.92318081 0.05340704]]
1.0
[[-0.40270429  0.97783197 -2.4286078  -1.03230615]
 [ 0.56923988 -0.35984316 -0.21824147 -0.87662429]
 [-0.16653558 -0.61798881  2.64684927  1.90893044]]
[  9.30108617   1.97824301 -11.27932918]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


The score for my Logistic Regression model, which is `1.0`, represents the model’s accuracy on the test set. This means that the model correctly predicted the class labels for 100% of the test samples. A perfect accuracy score suggests that the model is effectively separating the classes in this test set.

# Part 4: Support Vector Machine

In [103]:
# i. Use sklearn to train a Support Vector Classifier on the training set
from sklearn.svm import SVC
model = SVC(probability=True)
# ii. For a sample datapoint, predict the probabilities for each possible class
model.fit(X_train, y_train)
sample = X_test.iloc[0]
print(model.predict_proba([sample]))
print(model.predict([sample]))
# iii. Report on the score for the SVM, what does the score measure?
print(model.score(X_test, y_test))

[[0.01125184 0.97674606 0.01200209]]
['versicolor']
1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


The score I received for my Support Vector Classifier (SVC) model is `1.0`, which measures the model's accuracy on the test set. This score indicates that the model correctly predicted the class labels for 100% of the test samples. Additionally, when I tested the model with a sample data point, it predicted the label as `'versicolor'`, showing the specific class name. This perfect accuracy suggests that the model is effectively and accurately distinguishing between classes in this test set.

# Part 5: Neural Network

In [104]:
from sklearn.neural_network import MLPClassifier

# i. Train a Neural Network (MLP Classifier) on the training set
model = MLPClassifier()
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = X_test.iloc[0].values.reshape(1, -1)  # First sample in the test set
print("Predicted probabilities for each class (Default Model):", model.predict_proba(sample))
print("Predicted class (Default Model):", model.predict(sample))

# iii. Report on the score for the Neural Network, what does the score measure?
print("Default Model Score on Test Set:", model.score(X_test, y_test))

# iv: Experiment with different options for the neural network, report on your best configuration

# Variable to track the best model and score
best_score = 0
best_model = None

# Configuration 1: Default configuration
model1 = MLPClassifier()
model1.fit(X_train, y_train)
print("Predicted probabilities for each class (Model 1 - Default):", model1.predict_proba(sample))
score1 = model1.score(X_test, y_test)
print("Model 1 Score on Test Set:", score1)
if score1 > best_score:
    best_score = score1
    best_model = model1

# Configuration 2: Adding more neurons and layers with 'relu' activation and 'adam' solver
model2 = MLPClassifier(hidden_layer_sizes=(50, 50), activation='relu', solver='adam')
model2.fit(X_train, y_train)
print("Predicted probabilities for each class (Model 2 - Relu, Adam, 50x50):", model2.predict_proba(sample))
score2 = model2.score(X_test, y_test)
print("Model 2 Score on Test Set:", score2)
if score2 > best_score:
    best_score = score2
    best_model = model2

# Configuration 3: Adding more neurons and layers with 'tanh' activation and 'adam' solver
model3 = MLPClassifier(hidden_layer_sizes=(50, 50), activation='tanh', solver='adam', max_iter=1000)
model3.fit(X_train, y_train)
print("Predicted probabilities for each class (Model 3 - Tanh, Adam, 50x50):", model3.predict_proba(sample))
score3 = model3.score(X_test, y_test)
print("Model 3 Score on Test Set:", score3)
if score3 > best_score:
    best_score = score3
    best_model = model3

# Configuration 4: More neurons and layers with 'relu' activation and 'sgd' solver, lower learning rate
model4 = MLPClassifier(hidden_layer_sizes=(50, 50), activation='relu', solver='sgd', learning_rate_init=0.01, max_iter=1000)
model4.fit(X_train, y_train)
print("Predicted probabilities for each class (Model 4 - Relu, SGD, 50x50, LR=0.01):", model4.predict_proba(sample))
score4 = model4.score(X_test, y_test)
print("Model 4 Score on Test Set:", score4)
if score4 > best_score:
    best_score = score4
    best_model = model4

# Configuration 5: Experiment with deeper network
model5 = MLPClassifier(hidden_layer_sizes=(100, 100, 50), activation='relu', solver='adam', max_iter=1000)
model5.fit(X_train, y_train)
print("Predicted probabilities for each class (Model 5 - Relu, Adam, 100x100x50):", model5.predict_proba(sample))
score5 = model5.score(X_test, y_test)
print("Model 5 Score on Test Set:", score5)
if score5 > best_score:
    best_score = score5
    best_model = model5


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


Predicted probabilities for each class (Default Model): [[0.05495218 0.64584817 0.29919965]]
Predicted class (Default Model): ['versicolor']
Default Model Score on Test Set: 0.8666666666666667


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


Predicted probabilities for each class (Model 1 - Default): [[0.05426105 0.69967904 0.24605991]]
Model 1 Score on Test Set: 0.8666666666666667
Predicted probabilities for each class (Model 2 - Relu, Adam, 50x50): [[0.00692137 0.98046949 0.01260914]]
Model 2 Score on Test Set: 0.9333333333333333


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


Predicted probabilities for each class (Model 3 - Tanh, Adam, 50x50): [[0.00199394 0.99664916 0.0013569 ]]
Model 3 Score on Test Set: 1.0
Predicted probabilities for each class (Model 4 - Relu, SGD, 50x50, LR=0.01): [[0.00168804 0.99524109 0.00307087]]
Model 4 Score on Test Set: 0.9333333333333333
Predicted probabilities for each class (Model 5 - Relu, Adam, 100x100x50): [[8.76663616e-04 9.98797962e-01 3.25374559e-04]]
Model 5 Score on Test Set: 0.9333333333333333


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(



---

I trained multiple neural network models with various configurations, and the best-performing model achieved a perfect accuracy score of 1.0 (100%) on the test set. This top score was achieved by **Model 3**, configured with two hidden layers of 50 neurons each, using the `tanh` activation function and the `adam` solver, and trained for a maximum of 1000 iterations.

This configuration not only achieved perfect accuracy but also showed high confidence in predictions. For a sample data point, Model 3 classified it as `'versicolor'` with 99.66% confidence, indicating a reliable and robust prediction.

Based on these results, the optimal configuration is:
- **Two hidden layers with 50 neurons each (`hidden_layer_sizes=(50, 50)`), using the `tanh` activation function and the `adam` solver, with `max_iter=1000`**.

This setup provides the best balance of accuracy and confidence, making it the most effective model among those tested.

# Part 6: K-Nearest Neighbors

In [105]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier()
# ii. For a sample datapoint, predict the probabilities for each possible class
model.fit(X_train, y_train)
sample = X_test.iloc[0]
print(model.predict([sample]))
print(model.predict_proba([sample]))
# iii. Report on the score for kNN, what does the score measure?
print(model.score(X_test, y_test))

['versicolor']
[[0. 1. 0.]]
0.9333333333333333


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


The score for the k-Nearest Neighbors (kNN) model, which is `0.9333`, measures the model's accuracy on the test set. In the context of kNN, this score reflects how well the model can classify a sample by comparing it to its closest labeled data points in the training set. An accuracy of `0.9333` indicates that the model correctly classified 93.33% of the test samples based on the majority class of their nearest neighbors, showing that the kNN model effectively groups and identifies samples with similar feature patterns in the test data.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

### Best Performing Models

The **Logistic Regression, SVC, and Neural Network** models all achieved perfect accuracy on the test set, making them the best-performing models in this exercise. 

### Observations and Possible Explanations

The Logistic Regression and SVC models likely performed well due to the relatively linear separability of classes in the dataset, allowing these models to create effective decision boundaries. The Neural Network, despite being more complex, also performed optimally with the given configuration, suggesting that the dataset is small and structured enough to benefit from this more flexible model architecture.

The kNN model performed well but didn’t reach the same level of accuracy as the other models. Since kNN relies directly on the distance to neighboring points, it is slightly more sensitive to overlapping data points in feature space or variations in the test data.

### Surprising Aspects

One surprising aspect was that all models performed extremely well with minimal tuning, suggesting that the dataset may be inherently easy to classify due to clear distinctions between classes. Additionally, the Neural Network model reached high accuracy despite convergence warnings, indicating that the configurations were close to optimal even without extensive training. This exercise highlighted how different model types can effectively handle classification when the dataset is well-structured, even without complex tuning.